src/client_builder.rs

```diff
@@ -61,41 +155,132 @@ where
+    /// let client = ClientBuilder::new("imap.example.com", 993).native_tls()?;
+    /// # Ok(())
+    /// # }
+    /// ```
     #[cfg(feature = "native-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "native-tls")))]
-    pub fn native_tls(&mut self) -> Result<Client<TlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = TlsConnector::builder().build()?;
-            Ok(TlsConnector::connect(&ssl_conn, domain, tcp)?)
-        })
+    #[deprecated = "Use tls_kind(TlsKind::Native).connect() instead"]
+    pub fn native_tls(&mut self) -> Result<Client<Connection>> {
+        if self.mode == ConnectionMode::Tcp {
+            self.mode(ConnectionMode::Tls);
+        }
+
+        self.tls_kind(TlsKind::Native).connect()
     }
 
-    /// Return a new [`Client`] using `rustls` transport.
+    /// Makes a [`Client`] connection using the rustls TLS backend.
+    ///
+    /// Forces TLS unless [`ClientBuilder::starttls()`] was called
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
+    /// # {} #[cfg(feature = "rustls-tls")]
+    /// # fn main() -> Result<(), imap::Error> {
+    /// let client = ClientBuilder::new("imap.example.com", 993).rustls()?;
+    /// # Ok(())
+    /// # }
+    /// ```
     #[cfg(feature = "rustls-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "rustls-tls")))]
-    pub fn rustls(&mut self) -> Result<Client<RustlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = RustlsConnector::new_with_native_certs()?;
-            Ok(ssl_conn.connect(domain, tcp)?)
+    #[deprecated = "Use tls_kind(TlsKind::Rust).connect() instead"]
+    pub fn rustls(&mut self) -> Result<Client<Connection>> {
+        if self.mode == ConnectionMode::Tcp {
+            self.mode(ConnectionMode::Tls);
+        }
+
+        self.tls_kind(TlsKind::Rust).connect()
+    }
+
+    /// Make a [`Client`] using the configuration.
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
+    /// # {} #[cfg(feature = "rustls-tls")]
+    /// # fn main() -> Result<(), imap::Error> {
+    /// let client = ClientBuilder::new("imap.example.com", 143)
+    ///     .starttls().connect()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn connect(&self) -> Result<Client<Connection>> {
+        self.connect_with(|_domain, tcp| {
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            return self.build_tls_connection(tcp);
+            #[cfg(all(not(feature = "native-tls"), not(feature = "rustls-tls")))]
+            return Ok(tcp);
         })
     }
 
-    /// Make a [`Client`] using a custom TLS initialization. This function is intended
+    /// Make a [`Client`] using a custom initialization. This function is intended
     /// to be used if your TLS setup requires custom work such as adding private CAs
     /// or other specific TLS parameters.
     ///
+    /// Note, if the connection does not end up as TLS, handshake will not be called.
```

Comment:
This isn't _quite_ true because we may call `handshake` (currently at least) even if no TLS-enabling features are on, which would mean it's called despite the connection _not_ ending up as TLS.

Reply:
Which in another comment you suggested getting rid of connect_with entirely.  Do you still want to keep connect_with? (which I agree with the unreachable! comment before and will remove that code-path)